core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+    /// if the limit is reached.
+    fn schedulable_threads_with_read_only_handler(
+        &self,
+        account: &Pubkey,
+        read_only_handler: impl Fn(&ThreadSet, &[u32]) -> ThreadSet,
+    ) -> ThreadSet {
+        match (self.write_locks.get(account), self.read_locks.get(account)) {
+            (None, None) => ThreadSet::any(self.num_threads),
+            (None, Some((thread_set, counts))) => read_only_handler(thread_set, counts),
+            (Some((thread_id, count)), None) => {
+                if count == &self.sequential_queue_limit {
+                    ThreadSet::none()
+                } else {
+                    ThreadSet::only(*thread_id)
+                }
+            }
+            (Some((thread_id, count)), Some((thread_set, counts))) => {
+                debug_assert_eq!(Some(*thread_id), thread_set.only_one_scheduled());
+                if count + counts[*thread_id] == self.sequential_queue_limit {
+                    ThreadSet::none()
+                } else {
+                    ThreadSet::only(*thread_id)
+                }
+            }
+        }
+    }
+
+    /// Add locks for all writable and readable accounts on `thread_id`.
+    pub fn lock_accounts<'a>(
+        &mut self,
+        write_account_locks: impl Iterator<Item = &'a Pubkey>,
+        read_account_locks: impl Iterator<Item = &'a Pubkey>,
+        thread_id: ThreadId,
+    ) {
+        assert!(
+            thread_id < self.num_threads,
+            "thread_id must be < num_threads"
+        );
+        for account in write_account_locks {
+            self.write_lock_account(account, thread_id);
+        }
+
+        for account in read_account_locks {
+            self.read_lock_account(account, thread_id);
+        }
+    }
+
+    /// Locks the given `account` for writing on `thread_id`.
+    /// Panics if the account is already locked for writing on another thread.
+    fn write_lock_account(&mut self, account: &Pubkey, thread_id: ThreadId) {
+        match self.write_locks.entry(*account) {
+            Entry::Occupied(mut entry) => {
+                let (lock_thread_id, lock_count) = entry.get_mut();
+                assert_eq!(
+                    *lock_thread_id, thread_id,
+                    "outstanding write lock must be on same thread"
+                );
+
+                *lock_count += 1;
+                assert!(
+                    *lock_count <= self.sequential_queue_limit,
+                    "sequential queue limit must not be exceeded"
+                );
+            }
+            Entry::Vacant(entry) => {
+                entry.insert((thread_id, 1));
+            }
+        }
+
+        // Check for outstanding read-locks
+        if let Some((read_thread_set, _)) = self.read_locks.get(account) {
+            assert_eq!(
+                read_thread_set,
+                &ThreadSet::only(thread_id),
+                "outstanding read lock must be on same thread"
+            );
```

Comment:
nit: I'd prefer putting `&` outside business logic as much as possible like this:

```patch
         // Check for outstanding read-locks
-        if let Some((read_thread_set, _)) = self.read_locks.get(account) {
+        if let Some(&(read_thread_set, _)) = self.read_locks.get(account) {
             assert_eq!(
                 read_thread_set,
-                &ThreadSet::only(thread_id),
+                ThreadSet::only(thread_id),
                 "outstanding read lock must be on same thread"
             );
         }
```

Reply:
done: a5c692f2d5